### PR TITLE
Improve JSON column creation performance (list offsets)

### PR DESCRIPTION
**Pull request by karthikeyann**
## Description
In JSON parser during JSON column creation stage, col_ids is sorted twice - once inside reduce_to_column_tree and before list offset computation.
This PR avoids repeated sort by col_ids for list offset computation. Reuses previously sorted col_ids and node_ids.

Improves parsing bandwidth from 2.170GB/s to 2.414GB/s on GV100 for 1GB json string. (The parsing bandwidth computation includes json string transfer time. Excluding data transfer time, this PR improves parsing bandwidth from 3.9GB/s to 4.5GB/s)

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by karthikeyann** [comment]
# Benchmark Results Comparison

## nested_json_gpu_parser (old)

### [0] Quadro GV100

|    string_size    | Samples |  CPU Time  | Noise |  GPU Time  | Noise |  Elem/s  | bytes_per_second | peak_memory_usage |
|-------------------|---------|------------|-------|------------|-------|----------|------------------|-------------------|
|    2^20 = 1048576 |   2848x |   4.228 ms | 6.36% |   4.222 ms | 6.36% | 248.365M |        248365276 |        20.050 MiB |
|    2^21 = 2097152 |   1472x |   4.497 ms | 1.88% |   4.491 ms | 1.87% | 466.929M |        466928685 |        27.805 MiB |
|    2^22 = 4194304 |   1744x |   5.343 ms | 1.61% |   5.337 ms | 1.61% | 785.834M |        785833664 |        58.166 MiB |
|    2^23 = 8388608 |   2093x |   7.149 ms | 9.26% |   7.143 ms | 9.25% |   1.174G |       1174349893 |       124.409 MiB |
|   2^24 = 16777216 |     80x |  11.395 ms | 5.66% |  11.389 ms | 5.66% |   1.473G |       1473114329 |       140.173 MiB |
|   2^25 = 33554432 |    779x |  19.247 ms | 5.20% |  19.241 ms | 5.20% |   1.744G |       1743900305 |       353.410 MiB |
|   2^26 = 67108864 |    256x |  35.288 ms | 1.73% |  35.282 ms | 1.73% |   1.902G |       1902070564 |       596.186 MiB |
|  2^27 = 134217728 |    144x |  65.351 ms | 1.06% |  65.345 ms | 1.06% |   2.054G |       2053978434 |         1.118 GiB |
|  2^28 = 268435456 |    123x | 122.496 ms | 1.35% | 122.491 ms | 1.35% |   2.191G |       2191473166 |         2.218 GiB |
|  2^29 = 536870912 |     59x | 254.390 ms | 1.23% | 254.384 ms | 1.23% |   2.110G |       2110472128 |         4.331 GiB |
| 2^30 = 1073741824 |     31x | 494.872 ms | 0.61% | 494.868 ms | 0.61% |   2.170G |       2169753516 |         8.571 GiB |

## nested_json_gpu_parser (This PR)

### [0] Quadro GV100

|    string_size    | Samples |  CPU Time  | Noise |  GPU Time  | Noise |  Elem/s  | bytes_per_second | peak_memory_usage |
|-------------------|---------|------------|-------|------------|-------|----------|------------------|-------------------|
|    2^20 = 1048576 |   3120x |   4.086 ms | 8.67% |   4.081 ms | 8.67% | 256.963M |        256963187 |        21.114 MiB |
|    2^21 = 2097152 |   1440x |   4.310 ms | 1.98% |   4.304 ms | 1.97% | 487.261M |        487261194 |        27.578 MiB |
|    2^22 = 4194304 |    848x |   5.100 ms | 1.62% |   5.095 ms | 1.62% | 823.258M |        823257644 |        45.909 MiB |
|    2^23 = 8388608 |   2273x |   6.582 ms | 6.22% |   6.577 ms | 6.22% |   1.275G |       1275486550 |       129.244 MiB |
|   2^24 = 16777216 |    832x |  10.031 ms | 2.91% |  10.026 ms | 2.91% |   1.673G |       1673411309 |       180.019 MiB |
|   2^25 = 33554432 |    656x |  17.506 ms | 4.48% |  17.501 ms | 4.47% |   1.917G |       1917333507 |       340.467 MiB |
|   2^26 = 67108864 |    476x |  31.496 ms | 2.35% |  31.490 ms | 2.35% |   2.131G |       2131121538 |       642.379 MiB |
|  2^27 = 134217728 |    252x |  59.662 ms | 1.38% |  59.656 ms | 1.38% |   2.250G |       2249857835 |         1.162 GiB |
|  2^28 = 268435456 |    131x | 114.844 ms | 1.43% | 114.839 ms | 1.43% |   2.337G |       2337499645 |         2.225 GiB |
|  2^29 = 536870912 |     67x | 226.309 ms | 1.63% | 226.304 ms | 1.63% |   2.372G |       2372346897 |         4.344 GiB |
| 2^30 = 1073741824 |     34x | 444.692 ms | 0.83% | 444.726 ms | 0.83% |   2.414G |       2414389672 |         8.581 GiB |

**Review by ttnghia — APPROVED** [review]

File: cpp/src/io/json/json_column.cu [hunk 1]
```diff
@@ -333,9 +319,18 @@ void make_device_json_column(device_span<SymbolT const> input,
                              rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
+  rmm::device_uvector<NodeIndexT> original_col_ids(col_ids.size(), stream);  // make a copy
+  thrust::copy(rmm::exec_policy(stream), col_ids.begin(), col_ids.end(), original_col_ids.begin());
+
+  // sort_by_key {col_id}, {row_offset} stable
```
Comment (bdice) [inline cpp/src/io/json/json_column.cu:325]: This comment sounds like it's sorting by two key columns (`col_id, row_offset`), but I only see one, `col_ids`. What does `{row_offset}` mean in this comment?

File: cpp/src/io/json/json_column.cu [hunk 2]
```diff
@@ -466,46 +461,14 @@ void make_device_json_column(device_span<SymbolT const> input,
...
-          set_bit(d_columns_data[col_ids[i]].validity, row_offsets[i]);
-          d_columns_data[col_ids[i]].string_offsets[row_offsets[i]] = range_begin[i];
-          d_columns_data[col_ids[i]].string_lengths[row_offsets[i]] = range_end[i] - range_begin[i];
-          break;
-        default: break;
-      }
-    });
 
-  // 4. scatter List offset
+  // 3. scatter List offset
```
Comment (bdice) [inline cpp/src/io/json/json_column.cu:467]: I'd prefer to remove the numbers entirely. It's not necessary to track what step we're on.

File: cpp/src/io/json/json_column.cu [hunk 3]
```diff
@@ -541,9 +504,31 @@ void make_device_json_column(device_span<SymbolT const> input,
...
+     col_ids         = original_col_ids.begin(),
+     row_offsets     = row_offsets.begin(),
+     range_begin     = tree.node_range_begin.begin(),
+     range_end       = tree.node_range_end.begin(),
+     d_ignore_vals   = d_ignore_vals.begin(),
+     d_columns_data  = d_columns_data.begin()] __device__(size_type i) {
+      switch (node_categories[i]) {
+        case NC_STRUCT: set_bit(d_columns_data[col_ids[i]].validity, row_offsets[i]); break;
+        case NC_LIST: set_bit(d_columns_data[col_ids[i]].validity, row_offsets[i]); break;
+        case NC_VAL:
```
Comment (bdice) [inline cpp/src/io/json/json_column.cu:522]: Please use an explicit `[[fallthrough]]` any time that a `case` statement has no body and uses the subsequent `case`.
```suggestion
        case NC_VAL: [[fallthrough]];
```

**Review by bdice — CHANGES_REQUESTED** [review]

File: cpp/src/io/json/json_column.cu (hunk 1 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:325]: `{col_id}` is key, `{row_offset}` is value

File: cpp/src/io/json/json_column.cu (hunk 2 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:467]: Numbering helps specify the steps, and differentiate from other comments.

File: cpp/src/io/json/json_column.cu [hunk 4]
```diff
@@ -99,35 +99,36 @@ void print_tree(host_span<SymbolT const> input,
...
-  //   1. sort_by_key {col_id}, {row_offset} stable
-  rmm::device_uvector<NodeIndexT> node_ids(row_offsets.size(), stream);
-  thrust::sequence(rmm::exec_policy(stream), node_ids.begin(), node_ids.end());
-  thrust::stable_sort_by_key(rmm::exec_policy(stream),
-                             col_ids.begin(),
-                             col_ids.end(),
-                             thrust::make_zip_iterator(node_ids.begin(), row_offsets.begin()));
-  auto num_columns = thrust::unique_count(rmm::exec_policy(stream), col_ids.begin(), col_ids.end());
+  // 1. column count for allocation
+  auto num_columns =
```
Comment (bdice) [inline cpp/src/io/json/json_column.cu:120]: Can we add more `const` qualifiers?
```suggestion
  auto const num_columns =
```

**Review by bdice — APPROVED** [review]

**Comment by karthikeyann** [comment]
@gpucibot merge

**Pull request merged**